notebooks/03-query/15-pivot-unpivot.ipynb

```diff
@@ -0,0 +1,433 @@
+    "\n",
+    "A few cells aboce we saw that the values from PIVOT can be NULL. The query above calculates a ratio of human to total changes for each channel and the result is sorted on this ratio such that the channels with the highest proportion of robot updates are listed first. In order to calculate this metric even in presence of NULLs the query uses COALESCE on all the pivoted metrics.\n",
+    "\n",
+    "The result is much cleaner than the prior query and it is easy to determine which channels have the most robotic activity.\n",
+    "\n",
+    "Notice the use of SAFE_DIVIDE in the \"human_ratio\" calculation, which guards for division by zero for cases with no updates. Learn more about [SAFE_DIVIDE in the docs](https://druid.apache.org/docs/latest/querying/sql-functions#safe_divide).\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3f0af833-9508-4f26-a484-2778ca857a2a",
+   "metadata": {},
+   "source": [
+    "## Transform columns with UNPIVOT \n",
+    "UNPIVOT does the opposite of PIVOT. UNPIVOT turns columns into rows by using the names of the columns being removed as values for a new column called the \"names column\".\n",
+    "The values of the removed columns are combined into a single column called the \"values column\".\n",
+    "\n",
+    "Given the prior query you can investigate the sources of the updates in the most robotically updated pages.\n",
+    "\n",
+    "The following cell uses the UNPIVOT operation to do this:\n",
+    "```\n",
+    "  UNPIVOT ( \"changes\"  FOR \"action\" IN (\"added\", \"deleted\") )\n",
+    "```\n",
+    "- takes the values of multiple columns: `\"added\",\"deleted\"`\n",
+    "- incorporates them into a single column called the \"values column\": `\"changes\"`\n",
+    "- expands the results in \"names column\" with names of deleted columns: `\"action\"`\n",
+    "\n",
+    "Run the following cell to find the most active user as measured by total adding or deleting within the channel `'#it.wikipedia'` which has a ratio of ~0.50 (about half robots):"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f0159df8-6ca0-45af-9c68-6acd92defc4f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT \"user\",\n",
+    "       \"action\",\n",
+    "       SUM(\"changes\") \"total_changes\"\n",
+    "FROM \"example-wiki-pivot-unpivot\"\n",
+    "\n",
+    "UNPIVOT ( \"changes\"  FOR \"action\" IN (\"added\", \"deleted\") )\n",
+    "\n",
+    "WHERE TIME_IN_INTERVAL(__time, '2016-06-27/P1D')\n",
+    "       AND \"channel\"='#it.wikipedia' \n",
+    "       AND \"isRobot\"='true'\n",
+    "GROUP BY 1,2\n",
+    "ORDER BY 3 DESC \n",
+    "LIMIT 10\n",
+    "'''\n",
+    "display_client.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3e7f3f5d-bd0c-4957-8d62-94fb01def699",
+   "metadata": {},
+   "source": [
+    "The result provides the list of robot users that most updated the channel `'#it.wikipedia'`. \n",
```

Comment:
This sentence is a little confusing. Maybe something like this: "The query returns the list of robot users with the highest number of total updates..."

Reply:
Changed to :
 "The result provides the list of users that did the most additions or deletions in the channel `'#it.wikipedia'` and identified as `isRobot='true'`."